# Exploring Ebay Car Sales Data

The original data set was uploaded to [kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data), however for this project, DataQuest sampled 50,000 data points from the full dataset and dirtied the data to allow for data cleaning opportunities. Note the scraped data is from Ebay-Kleinanzeigen, so some of the content of the data is in German. A data dictionary is provided within Kaggle. 

## Goal of project
During this course, I have learned how to clean data using Pandas. I will apply what I have learned to clean the data and then analyze the used car listings. 




In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [4]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [5]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

After reviewing the data frame, I have noted that 15/20 columns are objects, however some of those can be cleaned and converted into integers including: 
 - price
 - odometer

Additionally, there are null values in the following columns which will need to be considered:
 - Vehicletype
 - gearbox
 - model
 - fuelType
 - notRepairedDamange

Note: Null values are not gerater than 20% of the column values for each column containing nulls
 
Data column names appear to be cleaned up as they are written in Camel Case instead of Snakecase and will require some modficiations. We are using integers for the index labels instead of a column name.

## Cleaning Columns

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos_copy = autos.copy()

In [8]:
## Changing the names of the columns to add clarity##
autos_copy.rename({"yearOfRegistration" : "registration_year", "monthOfRegistration" : "registration_month", "notRepairedDamage" : "unrepaired_damage", "dateCreated" : "ad_created"}, axis = 1, inplace = True)

In [9]:
## function from Geeks for Geeks that converts a string from Camelcase to Snakecase
def change_case(str): 
    res = [str[0].lower()] 
    for c in str[1:]: 
        if c in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'): 
            res.append('_') 
            res.append(c.lower()) 
        else: 
            res.append(c) 
      
    return ''.join(res) 

## loop over columns applying the above function and appending to copied dataframe
new_columns = []
for c in autos_copy.columns:
    clean_c = change_case(c)
    new_columns.append(clean_c)
    
autos_copy.columns = new_columns

In [10]:
autos_copy.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Above I viewed the columns, noting that they were all in [Camel Case](https://en.wikipedia.org/wiki/Camel_case) and felt that some column names were confusing. 

I first cleaned up the confusing column names and renamed them in snakecase. For the rest of the column names, I found a function from [Geeks for Geeks](https://www.geeksforgeeks.org/python-program-to-convert-camel-case-string-to-snake-case/) as I was not sure how to build a function that would covert strings from Camel Case to Snake Case. I was able to plug in the function and then applied the function to a loop over the column names. I then applied the modified column names to a copy of the autos data frame called autos_cols_mod. 

# Initial Exploration of Data and Cleaning

In [11]:
autos_copy.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Using the describe method I have noted the following:

1. The following columns have mostly one value that are candidates to be dropped and are not usesful for the analysis:
   - seller: with only 2 unique values, the top value is privat and occurrs 49,999 times
   - offer_type: only 2 unique values, the top value is Angebot and occurrs 49,999 times
   
2. The following columns will require more investigation: 
   - nr_of_pictures is an integer but there appears to be no values
   - I will check name to understand how this is displayed
   - vehicle_type - make sure there are no misspellings
   - resgitration_year appears to have errors with a min year of 1000 and a max year of 9999. (I will correct this later.) 
   - fuel_type to make sure no misspellings
   - brand for misspellings
   - postal codes - I am not sure why these are floats vs integers, I do not want to convert to integer however, in case rounding incorrectly changes the the postal code
 
3. The following columns contain data stored as text and will be converted and cleaned:
   - price
   - odometer
   
## 1. Removing columns

In [12]:
autos_copy["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

Because the column nr_of_pictures contains the value 0 for each row, it does not provide value to the anaysis and will be removed along with seller and offer_type

In [13]:
autos_copy.drop(["seller","offer_type","nr_of_pictures"], axis =1)

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price   abtest vehicle_type  registration_year    gearbox  power_p_s  \
0       $5,000  control          bus               2004    manuell        158   
1       $8,500  control    limousine               1997  automatik        286   
2       $8,990     test    limousine               2009    manuell        102   
3       $4,350  control   kleinwagen               2007  automatik         71   
4       $1,350     test        kombi               2003    manuell          0   
...        ...      ...          ...                ...        ...        ...   
49995  $24,900  control    limousine               2011  automatik        239   
49996   $1,980  control       cabrio               1996    manuell         75   
49997  $13,200     test       cabrio               2014  automatik         69   
49998  $22,900  control        kombi               2013    manuell        150   
49999   $1,250  control    limousine               1996    manuell        101   

        model   odometer  registration_month fuel_type       brand  \
0      andere  150,000km                   3       lpg     peugeot   
1         7er  150,000km                   6    benzin         bmw   
2        golf   70,000km                   7    benzin  volkswagen   
3      fortwo   70,000km                   6    benzin       smart   
4       focus  150,000km                   7    benzin        ford   
...       ...        ...                 ...       ...         ...   
49995      q5  100,000km                   1    diesel        audi   
49996   astra  150,000km                   5    benzin        opel   
49997     500    5,000km                  11    benzin        fiat   
49998      a3   40,000km                  11    diesel        audi   
49999  vectra  150,000km                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

In [14]:
autos_copy["name"].unique()

array(['Peugeot_807_160_NAVTECH_ON_BOARD',
       'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik',
       'Volkswagen_Golf_1.6_United', ...,
       'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon',
       'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!',
       'Fiat_500_C_1.2_Dualogic_Lounge'], dtype=object)

## 2. Investigating additional columns
There does not appear to be a particular naming convention for values within the name column. Given there are 38,754 unique values, I will leave name as is. Next I will review vehicle type, model, and brand

## Check for any misspellings prior to aggregation

In [15]:
autos_copy["vehicle_type"].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [16]:
autos_copy["vehicle_type"].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

Although some values are in German, there does not appear to be any misspelled words that would result in aggregation errors. However, after performing a google search, I did discover that andere means other. And as discussed above, this column contains 5,095 Null (nan) values or 10.2%. Because this is below 20% of the total values, I will leave the Null values as is. 

In [17]:
autos_copy["fuel_type"].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [18]:
autos_copy["fuel_type"].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

Similarly, for fuel types,there does not appear to be any misspelled words that would result in aggregation errors. There are 22 "andere" or other values, which is only 0.04% of the column values. This column contains 4.482 Null (nan) values or 8.96%. Because this is below 20% of the total values, I will leave the Null values as is. 

In [19]:
autos_copy["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [20]:
autos_copy["brand"].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
citroen             701
smart               701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

Brand does not contain any null values and there do not appear to be any misspellings that would cause aggregation errors. 

## Converting Price and Odometer Columns from string to numeric values

In [21]:
autos_copy["price"] = autos_copy["price"].str.replace("$","").str.replace(",","").astype(int)
autos_copy.rename({"price":"price_dollars"}, axis=1, inplace=True)

In [22]:
## Somehow accidentally ran this, did the conversion, and then ran into an error not
## not realizing the conversion had occurred. I deleted the code so here it is in comments
autos_copy["odometer"] = autos_copy["odometer"].str.replace(",", "").str.replace("km","").astype(int)
autos_copy.rename({"odometer": "odometer_km"}, axis =1, inplace = True)
autos_copy["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [23]:
autos_copy.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_dollars   abtest vehicle_type  registration_year  \
0  privat    Angebot           5000  control          bus               2004   
1  privat    Angebot           8500  control    limousine               1997   
2  privat    Angebot           8990     test    limousine               2009   
3  privat    Angebot           4350  control   kleinwagen               2007   
4  privat    Angebot           1350     test        kombi               2003   

     gearbox  power_p_s   model  odometer_km  registration_month fuel_type  \
0    manuell        158  andere       150000                   3       lpg   
1  automatik        286     7er       150000                   6    benzin   
2    manuell        102    golf        70000                   7    benzin   
3  automatik         71  fortwo        70000                   6    benzin   
4    manuell          0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Exploring the Price and Odometer Columns for Outliers

First I will look for unique values within each column, then use the describe method to view statistics, sort the values, and remove outliers

## Price Column

In [24]:
unique_price = autos_copy["price_dollars"].unique().shape
print(unique_price)

(2357,)


In [25]:
price_stats = autos_copy["price_dollars"].describe()
print(price_stats)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64


Reviewing the above, I have noted that there are 2357 unique prices. The mean price is 9,840.04 and the min price is 0 and the max is 100,000,000. 

In [26]:
autos_copy["price_dollars"].value_counts().head(20)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price_dollars, dtype: int64

Looking at the value counts, the prices appear to be rounded. I am unsure whether the site rounds numbers, this has been cleaned by Kaggle, or users tend to input prices into rounded numbers to compete with other sellers. There are 1,421 values with a 0 dollar price, which is 2.8% of the car values. The $0 car rows may therefore be removed. 

I would like to futher explore the max price of one hundred million dollars. This seems high, so I will first sort the price column from highest to lowest and use value counts to get an idea of whether this is an outlier or not

In [27]:
##price column sorted from highest to lowest
autos_copy["price_dollars"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_dollars, dtype: int64

In [28]:
##price column sorted from lowest to highest
autos_copy["price_dollars"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_dollars, dtype: int64

There are about 14 car prices over 1 million, representing 0.03% of the prices. I also noticed that prices jump from 999,990 down to 350,000 a difference of 649,990 which is the largest price variance. Because of that, prices over 350,000 can be removed as outliers.

While there are about 250 cars priced between 1 and 35, this is likely okay because Ebay is a bidding site and occassionally bids do start at 1. However, the 0 values will be removed as outliers. 

To remove the outliers, I will use between to assign the values between 1 and 350000 (used 351000 to ensure 350000 is included). I had tried to use the df.drop method but unfortunately did not do it correctly and delete all my rows. I see how this method below is cleaner to read and simply just assigns the values meeting the condition. Lesson learned!

In [29]:
autos_copy = autos_copy[autos_copy["price_dollars"].between(1,351000)]

In [30]:
print(autos_copy["price_dollars"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dollars, dtype: float64


Through that I removed 1,435 values, most of those were 0. My mean price is now 5,888 with a max of 350,000 and a min of 1 dollar. 

# Odometer Column

In [31]:
autos_copy["odometer_km"].unique().shape

(13,)

In [32]:
autos_copy["odometer_km"].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [33]:
autos_copy["odometer_km"].value_counts().head(13)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

Odometers, compared to price seem to be more evenly distributed without any outliers. Again, the values appear to be rounded, which I think may be either by the site or the cleaning done by Kaggle. I find it unlikely that all cars have even odometer readings. However, because the range makes sense, no outliers will be dropped. 

Next, I will explore and clean up the date and time columns: date_crawled, last_seen, ad_created, registration_month, registration_year. 

# Exploring and Cleaning Date Columns

I will first look at the date crawled, ad created, and last seen columns as they all follow the same YYYY-MM-DD HH:MM:SS format and are currently all contained as string values. I will begin by exploring the date range for each column by first extracting the dates (first 10 characters), obtain the value counts and sort the dates. 

In [34]:
#date_crawled
autos_copy["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

## Date Crawled
This column was added by the crawler and covers the dates 3/5/16 - 4/7/16. The crawler was run daily during this time period. There is not one day that appears to have significantly more listings compared to the others, rather the distribution is about uniform across the days. 

In [35]:
## ad_created
autos_copy["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

## Ad Created
This column came from the website and contains 76 total value. The Ad created dates range from 6/11/2015 - 4/7/2016, making the oldest ad about 9 months old. There are a higher percentage of ads fall 1-2 months within the listing date of 3/5/16 - 4/17/16 identified above, however there are older ads. 

In [36]:
## last_seen
autos_copy["last_seen"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

## Last Seen 
This column is also created by the crawler and contains the same date ranges as date_crawled - 3/5/16 - 4/7/16. The last seen date allows to understand either when the car was sold or when the crawling date ended. Given the last 3 days in the period saw a significant increase, this is indicative the last seen shows the end of a crawling period vs a car sale. 

# Registration Year

In [37]:
autos_copy["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

It appears there may be some incorrect data in the registration year as the max year is 9999 (too far into the future and min year is 1000 (before cars were invented). 

Because the data set is for March and April of 2016, a registration year cannot be after 2016. However, the first patent was issued in 1886, meaning that early 1900 dates are possible. 

I will search for the number of car listing that fall outside of the 1886-2016 interval first and then determine how to handle the values ie delete rows or replace values.

In [38]:
autos_copy["registration_year"].between(1886, 2016).value_counts(normalize = True)

True     0.961207
False    0.038793
Name: registration_year, dtype: float64

Because only 4% of registration years fall outside the range of 1886-2016, I can comfortably delete those rows without disrupting the results of my analysis. 

In [39]:
autos_copy = autos_copy[autos_copy["registration_year"].between(1886, 2016)]
autos_copy["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

# Exploring Price by Brand
Now that the data has been cleaned, I will next begin to aggregate based on the brand column. 

In [49]:
brand_counts = autos_copy["brand"].value_counts()
print(brand_counts)
len(brand_counts)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64


40

In [58]:
brand_percentage = autos_copy["brand"].value_counts(normalize = True)
print(brand_percentage)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

## Summary of Brands
There are 40 different car brands listed for the time period. German car brands (volkswagen, bms, opel, mercedes_benz, and audi) make up the top 5 most popular brands with volkswagen being the most popular brand. The number of volkswagens is nearly double the number of bmws, the second most popular brand. 

Becuase there are many brands that make up a small percentage of the listings, I will focus on brands with 1000+ listings (2.5 percent or more) to get an idea of the average prices per most popular brands. 

In [60]:
brand_percentage = autos_copy["brand"].value_counts(normalize = True)
popular_brands = brand_percentage[brand_percentage > .025].index
print(popular_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat'],
      dtype='object')


In [64]:
brand_mean_price = {}

for brand in popular_brands:
    selected_brand = autos_copy[autos_copy["brand"] == brand]
    mean_price = selected_brand["price_dollars"].mean()
    brand_mean_price[brand] = int(mean_price)
    
brand_mean_price

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813}

Of the top 9 brands, the following are greater than 5000 dollars:
 - audi
 - mercedes_benz
 - bmw
 
 The following are less than 5000 dollars:
  - ford
  - peugeot
  - fiat
  - opel 
  
Volkwagon, the most popular brand is priced at 5402 which may make it a popular listing - it is not too pricey for buyers but not too cheap for sellers. 


## Odometer readings for most popular brands

In [65]:
brand_mean_mileage = {}

for brand in popular_brands:
    selected_brand = autos_copy[autos_copy["brand"] == brand]
    mean_mileage = selected_brand["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
brand_mean_price

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813}

To search for any similiarties between price and mileage, I will next convert the two dicitionaries 1) mean price and 2) mean mileage into series objects using the series constructor. I will then combine them into a dataframe using the dataframe constructuer to make a easy to visualize comparision. 

In [66]:
bmp_series = pd.Series(brand_mean_price)
bmm_series = pd.Series(brand_mean_mileage)
print(bmp_series)
print(bmm_series)

volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
renault          2474
peugeot          3094
fiat             2813
dtype: int64
volkswagen       128707
bmw              132572
opel             129310
mercedes_benz    130788
audi             129157
ford             124266
renault          128071
peugeot          127153
fiat             117121
dtype: int64


In [69]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])

In [70]:
df['mean_mileage'] = bmm_series

In [71]:
print(df)

               mean_price  mean_mileage
volkswagen           5402        128707
bmw                  8332        132572
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9336        129157
ford                 3749        124266
renault              2474        128071
peugeot              3094        127153
fiat                 2813        117121


There does not appeart to be a correlation between price and mileage. All cars have mileage over 100k km and vary only by about 15k km. However, there is a slight trend where the more expensive cars have a slightly higher mileage. This could be due to the brand name rather than the desire for more mileage however. 